### DISCUSSION TITLE
how to debug git-lfs source code

### DISCUSSION THREAD

**Discussion by aylwynlake** [git-lfs/git-lfs#5691] [Q&A] [answered]
When I execute the `git clone` command, is there a way to debug Git LFS source code within Goland? Alternatively, how can I retrieve the parameters used when running the git-lfs command during `git clone`? I’d like to start debugging from the entry point in git-lfs.go. The reason for doing this is that I want to achieve the following functionality:

I intend to develop a custom transfer agent. My requirement is as follows: when I execute `git push`, LFS should upload to the LFS server store configured for the Git repository service. However, when I execute `git clone` or `git pull`, LFS should first attempt to download from an internal cache address. If it fails to find the file in the cache, it should then download from the LFS server store provided by the Git repository service. I’ve looked at this project as a reference: [git-lfs-agent-scp.sh](https://github.com/yoshimoto/git-lfs-agent-scp.sh), but I’m still unsure how to implement this functionality.

**Comment by chrisd8088** [comment, marked as answer]
As far as analyzing the sequence of events during a `git clone`, you might try using the `GIT_TRACE=1` environment variable (i.e., `GIT_TRACE=1 git clone`, if you're using Linux or macOS).  The `GIT_TRANSFER_TRACE=1` and `GIT_CURL_VERBOSE=1` variables can also provide additional information.

What you are likely to encounter is that with Git LFS configured, Git will run the `git lfs filter-process` command to "smudge" the blobs as they are populated in your working tree, meaning download the Git LFS objects corresponding to any Git blobs which are Git LFS pointers.  You can find the source code for that command in the `commands/command_filter_process.go` [file](https://github.com/git-lfs/git-lfs/blob/c4dc541ab5f5d819b27d8ae839ae932a3c9dce93/commands/command_filter_process.go).

Before you go to the trouble of implementing anything in code, though, you might investigate whether the `lfs.pushurl` or `remote.<remote>.lfspushurl` Git configuration settings might provide the functionality you need.  The Git LFS client can be configured to connect to a different URL when pushing than when fetching, so you could direct pushes to the server, while for fetches the client could connect to a proxy which first checks its cache and then fetches from the server.  You'd still have to implement the proxy, but you wouldn't need to modify the Git LFS client itself.

**Reply by aylwynlake** [reply]
Thank you so much for clarifying. Based on your response, I’ve located this issue: [Endpoints for upload and download](https://github.com/git-lfs/git-lfs/issues/2367) similar to my question.

**Comment by aylwynlake** [comment]
for someone who want to debug git-lfs code, it's very simple.
![debug](https://github.com/git-lfs/git-lfs/assets/1916613/b2cec1b7-86ac-421f-941c-afb3f4bdbcc7)
